packages/core/echo/echo-pipeline/src/automerge/collection-synchronizer.ts

```diff
@@ -0,0 +1,197 @@
+//
+// Copyright 2024 DXOS.org
+//
+
+import { Event, scheduleTask, scheduleTaskInterval } from '@dxos/async';
+import { next as am } from '@dxos/automerge/automerge';
+import type { DocumentId, PeerId } from '@dxos/automerge/automerge-repo';
+import { Resource, type Context } from '@dxos/context';
+import { defaultMap } from '@dxos/util';
+
+const MIN_QUERY_INTERVAL = 5_000;
+
+const POLL_INTERVAL = 30_000;
+
+export type CollectionSynchronizerParams = {
+  sendCollectionState: (collectionId: string, peerId: PeerId, state: CollectionState) => void;
+  queryCollectionState: (collectionId: string, peerId: PeerId) => void;
+  shouldSyncCollection: (collectionId: string, peerId: PeerId) => Promise<boolean>;
+};
+
+/**
+ * Implements collection sync protocol.
+ */
+export class CollectionSynchronizer extends Resource {
+  private readonly _sendCollectionState: CollectionSynchronizerParams['sendCollectionState'];
+  private readonly _queryCollectionState: CollectionSynchronizerParams['queryCollectionState'];
+  private readonly _shouldSyncCollection: CollectionSynchronizerParams['shouldSyncCollection'];
+
+  /**
+   * CollectionId -> State.
+   */
+  private readonly _perCollectionStates = new Map<string, PerCollectionState>();
+
+  private readonly _connectedPeers = new Set<PeerId>();
+
+  public readonly remoteStateUpdated = new Event<{ collectionId: string; peerId: PeerId }>();
+
+  constructor(params: CollectionSynchronizerParams) {
+    super();
+    this._sendCollectionState = params.sendCollectionState;
+    this._queryCollectionState = params.queryCollectionState;
+    this._shouldSyncCollection = params.shouldSyncCollection;
+  }
+
+  protected override async _open(ctx: Context): Promise<void> {
+    scheduleTaskInterval(
+      this._ctx,
+      async () => {
+        for (const collectionId of this._perCollectionStates.keys()) {
+          this.refreshCollection(collectionId);
```

Comment:
maybe async return after refreshCollection to yield cpu time to other tasks

Reply:
to add jitter for collection queries?

Reply:
I think there should be kind of implicit jitter cause of MIN_QUERY_INTERVAL which is nice. async return is just for spreading iteration over all_document_count * peers over multiple callbacks iterating over document_per_collection * peers in each